lib/libjenkinspillage.py

```diff
@@ -12,109 +14,171 @@
 #     'http': 'socks4a://localhost:9080',
 #     'https': 'socks4a://localhost:9080'
 
-class JenkinsJobBuild():
+
+class JenkinsConnection:
     """Class to interact with builds once you find them"""
+
     def __init__(self, url, proxies={}, auth={}):
         # validate the URL before using it
         url_validation = urllib.parse.urlparse(url)
-        if not all([url_validation.scheme , url_validation.netloc , url_validation.path]):
+        if not all([url_validation.scheme, url_validation.netloc, url_validation.path]):
             raise ValueError
         self.url = url
-        self.base_url = url_validation.scheme + '://' + url_validation.netloc
+        self.base_url = url_validation.scheme + "://" + url_validation.netloc
         self.auth = auth
-        # if "hudson.model.FreeStyleBuild" not in self._api_request(url):
+        # if "hudson.model.FreeStyleBuild" not in self._api_get(url):
         #     print("You did not specify a correct build URL: {}".format(url))
         #     raise ValueError
-    
-    def _api_request(self, url):
-        return requests.get(url, verify=False, auth=self.auth).text
+
+    def _add_crumb_header(self, url: str) -> dict:
+        """
+        Add crumb header to request.  Largely adopted from openstack/python-jenkins
+        """
+
+        response = requests.get(
+            f"{url}/crumbIssuer/api/json", auth=self.auth, verify=False
+        )
+
+        if response.ok:
+            crumb = response.json()
+            return {f"{crumb['crumbRequestField']}": f"{crumb['crumb']}"}
+
+        return {}
+
+    def _api_get(self, url: str) -> str:
+        return requests.get(url, verify=False, auth=self.auth)
+
+    def _api_post(self, url: str, data: dict = None) -> str:
+        return requests.post(
+            url,
+            verify=False,
+            auth=self.auth,
+            data=data,
+            headers=self._add_crumb_header(url=self.url),
+        )
+
+    def execute_script(self, script):
+        """
+        Open Groovy script and send to
```

Comment:
"send to server" -  I assume you meant to say?

Reply:
I'll update the docstring in the coming commit.